pkg/peer/webrtc.go

```diff
@@ -0,0 +1,171 @@
+				p.logger.Errorf("Failed to send RTCP PLI: %v", err)
+				return
+			}
+		}
+	}()
+
+	// Create a local track, all our SFU clients that are subscribed to this
+	// peer (publisher) wil be fed via this track.
+	localTrack, err := webrtc.NewTrackLocalStaticRTP(
+		remoteTrack.Codec().RTPCodecCapability,
+		remoteTrack.ID(),
+		remoteTrack.StreamID(),
+	)
+	if err != nil {
+		p.logger.WithError(err).Error("failed to create local track")
+		return
+	}
+
+	// Notify others that our track has just been published.
+	p.sink.Send(NewTrackPublished{Track: localTrack})
+
+	// Start forwarding the data from the remote track to the local track,
+	// so that everyone who is subscribed to this track will receive the data.
+	go func() {
+		rtpBuf := make([]byte, 1400)
+
+		for {
+			index, _, readErr := remoteTrack.Read(rtpBuf)
+			if readErr != nil {
+				if readErr == io.EOF { // finished, no more data, no error, inform others
+					p.logger.Info("remote track closed")
+				} else { // finished, no more data, but with error, inform others
+					p.logger.WithError(readErr).Error("failed to read from remote track")
+				}
+				p.sink.Send(PublishedTrackFailed{Track: localTrack})
+				return
+			}
+
+			// ErrClosedPipe means we don't have any subscribers, this is ok if no peers have connected yet.
+			if _, err = localTrack.Write(rtpBuf[:index]); err != nil && !errors.Is(err, io.ErrClosedPipe) {
+				p.logger.WithError(err).Error("failed to write to local track")
+				p.sink.Send(PublishedTrackFailed{Track: localTrack})
+				return
+			}
+		}
+	}()
+}
+
+// A callback that is called once we receive an ICE candidate for this peer connection.
+func (p *Peer[ID]) onICECandidateGathered(candidate *webrtc.ICECandidate) {
+	if candidate == nil {
+		p.logger.Info("ICE candidate gathering finished")
+		p.sink.Send(ICEGatheringComplete{})
+		return
+	}
+
+	p.logger.WithField("candidate", candidate).Debug("ICE candidate gathered")
+	p.sink.Send(NewICECandidate{Candidate: candidate})
+}
+
+// A callback that is called once we receive an ICE connection state change for this peer connection.
+func (p *Peer[ID]) onNegotiationNeeded() {
+	p.logger.Debug("negotiation needed")
+	offer, err := p.peerConnection.CreateOffer(nil)
+	if err != nil {
+		p.logger.WithError(err).Error("failed to create offer")
+		return
+	}
+
+	if err := p.peerConnection.SetLocalDescription(offer); err != nil {
+		p.logger.WithError(err).Error("failed to set local description")
+		return
+	}
+
+	p.sink.Send(RenegotiationRequired{Offer: &offer})
+}
+
+// A callback that is called once we receive an ICE connection state change for this peer connection.
+func (p *Peer[ID]) onICEConnectionStateChanged(state webrtc.ICEConnectionState) {
+	p.logger.WithField("state", state).Info("ICE connection state changed")
+
+	switch state {
+	case webrtc.ICEConnectionStateFailed, webrtc.ICEConnectionStateDisconnected:
+		// TODO: Ask Simon if we should do it here as in the previous implementation.
```

Comment:
The previous implementation doesn't handle this afaik, so we don't have to worry about this for now I think